File: cpp/poker/table_output.cpp
#include<cstdio>
//#include "deck.cpp"


struct card {
  char value;
  char naipe;
};

struct player {
  card hand[2];
  int chips;
  int lastBet;
  char role;
  float preFlopProb; // 2 cartas
  float flopToTurnProb; // 5 cartas
  float turnToRiverProb; // 6 cartas 
  float riverToShowDownProb; // 7 cartas
  //string action;
  bool showCards;
  bool active;
};

int numDigits(int num){
    int digits = 1;
    int aux = 10;
    while((num / aux) >= 1){
        digits++;
        aux = aux * 10;
    }
    return digits;
}

void topBorder(){
    //printf("%c", 218);
    printf("┌");
    
    for(int i = 0; i < 90; i++){
      //  printf("%c", 196);
        printf("─");
    }
    //printf("%c\n",191);
    printf("┐\n");
}

void bottomBorder(){
    //printf("%c", 192);
    printf("└");
    
    for(int i = 0; i < 12; i++){
        printf("─");
    }
    printf("┴");
    for(int i = 0; i < 77; i++){
        printf("─");
    }
    //printf("%c\n",217);
    printf("┘\n");

}

void lateralBorder(int n){
    for(int i = 0; i < n; i++){
        //printf("%c\n", 179);       \n   
        printf("│");
        for(int j = 0; j < 90; j++){
            printf(" ");
        }
        //printf("%c\n", 179);     
        printf("│\n");
    }

}

void spaces(int n){
    for(int j = 0; j < n; j++){
            printf(" ");
    }
}

void centralSpaces(){
    spaces(64);
}

void centralCardSpaces(){
    spaces(40);
}

void lateralSpaces(){
    spaces(3);
}

void cardTop(){
    //printf("%c", 218);  
    printf("┌");
    for(int j = 0; j < 3; j++){
        //printf("%c", 196);
        printf("─");
    }
    //printf("%c",191);
    printf("┐");
}

void cardTop(int n){
    for(int i = 0; i < n; i++){
        cardTop();
    }
}

void cardLateral(){
    //printf("%c\n", 179);      
    printf("│");
        for(int j = 0; j < 3; j++){
            printf(" ");
        }
    //printf("%c\n", 179);      
    printf("│");
    
}

void cardLateral(int n){
    for(int i = 0; i < n; i++){
        cardLateral();
    }   
}

void cardLateral(char c){   
    //printf("%c\n", 179);      
    printf("│");
    
    if(c == 'T'){
        printf("10 ");      
    }else{
        printf("%c  ", c);
    }
    //printf("%c\n", 179);      
    printf("│");  
}

void cardBottom(){
    //printf("%c", 192);
    printf("└");  
    for(int j = 0; j < 3; j++){
        //printf("%c", 196);
        printf("─");
    }
    //printf("%c\n",217);
    printf("┘");

}

void cardBottom(int n){
    for(int i = 0; i < n; i++){
        cardBottom();
    }
}

void centralCard(player p){      
    printf("│");

    centralCardSpaces();
    cardTop(2);
    centralCardSpaces();
       
    printf("│\n");
         
    printf("│");

    centralCardSpaces();
    char value1 = ' ';
    char value2 = ' ';
    if(p.showCards){
        value1 = p.hand[0].value;
        value2 = p.hand[1].value;
    }
    cardLateral(value1);
    cardLateral(value2);
    centralCardSpaces();
;     
    printf("│\n");     
   
    printf("│");

    centralCardSpaces();
    char naipe1 = ' ';
    char naipe2 = ' ';
    if(p.showCards){
        naipe1 = p.hand[0].naipe;
        naipe2 = p.hand[1].naipe;
    }
    cardLateral(naipe1);
    cardLateral(naipe2);
    centralCardSpaces();
  
    printf("│\n");     
     
    printf("│");

    centralCardSpaces();
    cardBottom(2);
    centralCardSpaces();
     
    printf("│\n");
}

void centralCardWithProb(player p, float prob){
    printf("│");

    centralCardSpaces();
    cardTop(2);
    centralCardSpaces();
       
    printf("│\n");

    printf("│");     
    
    centralCardSpaces();
    cardLateral(p.hand[0].value);
    cardLateral(p.hand[1].value);
    centralCardSpaces();
;     
    printf("│\n");

    printf("├");

    for(int i = 0; i < 12; i++)
        printf("─");
    printf("┐");

    
    spaces(27);
    cardLateral(p.hand[0].naipe);
    cardLateral(p.hand[1].naipe);
    centralCardSpaces();

    printf("│\n");

    printf("│");
    char c = '%';
    printf(" WIN: %.1f", prob);
    printf("%c", c);
    if(numDigits(prob) >= 1){
        spaces(3 - numDigits(prob));
    }
    printf("│");

    spaces(27);
    cardBottom(2);
    centralCardSpaces();

    printf("│\n");
}


void cardsLateral(player p1, player p2){//card c1, card c2, card c3, card c4){
    //printf("%c\n", 179);
    printf("│");
    lateralSpaces();
    cardTop(2);
    centralSpaces();
    cardTop(2);
    lateralSpaces();
    //printf("%c\n", 179);
    printf("│\n");

    //printf("%c\n", 179);
    printf("│");
    lateralSpaces();
    char value1 = ' ';
    char value2 = ' ';
    if(p1.showCards){
        value1 = p1.hand[0].value;
        value2 = p1.hand[1].value;
    }
    cardLateral(value1);
    cardLateral(value2);
    centralSpaces();

    value1 = ' ';
    value2 = ' ';
    if(p2.showCards){
        value1 = p2.hand[0].value;
        value2 = p2.hand[1].value;
    }
    cardLateral(value1);
    cardLateral(value2);
    lateralSpaces();
    //printf("%c\n", 179);
    printf("│\n");

    //printf("%c\n", 179);
    printf("│");
    lateralSpaces();
    char naipe1 = ' ';
    char naipe2 = ' ';
    if(p1.showCards){
        naipe1 = p1.hand[0].naipe;
        naipe2 = p1.hand[1].naipe;
    }
    cardLateral(naipe1);
    cardLateral(naipe2);

    centralSpaces();

    naipe1 = ' ';
    naipe2 = ' ';
    if(p2.showCards){
        naipe1 = p2.hand[0].naipe;
        naipe2 = p2.hand[1].naipe;
    }
    cardLateral(naipe1);
    cardLateral(naipe2);
    lateralSpaces();
    //printf("%c\n", 179);
    printf("│\n");

    //printf("%c\n", 179);
    printf("│");
    lateralSpaces();
    cardBottom(2);
    centralSpaces();
    cardBottom(2);
    lateralSpaces();
    //printf("%c\n", 179);
    printf("│\n");
}

void flopTurnRiver(card cards[]){
    //printf("%c\n", 179);
    printf("│");
    spaces(30);
    cardTop(3);
    spaces(2);
    cardTop();
    spaces(2);
    cardTop();
    spaces(31);
    //printf("%c\n", 179);
    printf("│\n");

    //printf("%c\n", 179);
    printf("│");
    spaces(30);
    cardLateral(cards[0].value);
    cardLateral(cards[1].value);
    cardLateral(cards[2].value);
    spaces(2);
    cardLateral(cards[3].value);
    spaces(2);
    cardLateral(cards[4].value);
    spaces(31);
    //printf("%c\n", 179);
    printf("│\n");

    //printf("%c\n", 179);
    printf("│");
    spaces(30);
    cardLateral(cards[0].naipe);
    cardLateral(cards[1].naipe);
    cardLateral(cards[2].naipe);
    spaces(2);
    cardLateral(cards[3].naipe);
    spaces(2);
    cardLateral(cards[4].naipe);
    spaces(31);
    //printf("%c\n", 179);
    printf("│\n");

    //printf("%c\n", 179);
    printf("│");
    spaces(30);
    cardBottom(3);
    spaces(2);
    cardBottom();
    spaces(2);
    cardBottom();
    spaces(31);
    //printf("%c\n", 179);
    printf("│\n");

}

void printCentralPlayer(player p, int numPlayer){
    //printf("%c\n", 179);
    printf("│");
    spaces(39);
    if(!p.active)
        printf("X");
    else
        spaces(1);
    printf("Player %d", numPlayer);
    int numSpaces = 42;
    if(p.role != 'C'){
        printf("(%c)", p.role);
        numSpaces = numSpaces - 3;
    }
    spaces(numSpaces);
    //printf("%c\n", 179);
    printf("│\n");

    int digits = numDigits(p.chips);
    //printf("%c\n", 179);
    printf("│");
    spaces(40);
    printf("Chips: %d", p.chips);
    spaces(43 - digits);
    //printf("%c\n", 179);
    printf("│\n");

    digits = numDigits(p.lastBet);
    printf("│");
    spaces(40);
    printf("Bet: %d", p.lastBet);
    spaces(45 - digits);
    //printf("%c\n", 179);
    printf("│\n");


}

void printLateralPlayers(player p1, int numPlayer1, player p2, int numPlayer2){
    //printf("%c\n", 179);
    printf("│");
    if(!p1.active){
        spaces(2);
        printf("X");
    }else
        lateralSpaces();
    printf("Player %d", numPlayer1);
    int numSpaces = 65;
    if(p1.role != 'C'){
        printf("(%c)", p1.role);
        numSpaces = numSpaces - 3;
    }
    spaces(numSpaces);
    if(p2.role != 'C'){
        printf("(%c)", p2.role);
    }else{
        spaces(3);
    }
    printf("Player %d", numPlayer2);
    if(!p2.active){
        printf("X");
        spaces(2);
    }else
        lateralSpaces();
    //printf("%c\n", 179);
    printf("│\n");

    //printf("%c\n", 179);
    printf("│");
    lateralSpaces();
    printf("Chips: %d", p1.chips);
    spaces(70 - numDigits(p1.chips) - numDigits(p2.chips));
    printf("Chips: %d", p2.chips);
    lateralSpaces();
    //printf("%c\n", 179);
    printf("│\n");

    printf("│");
    lateralSpaces();
    printf("Bet: %d", p1.lastBet);
    spaces(74 - numDigits(p1.lastBet) - numDigits(p2.lastBet));
    printf("Bet: %d", p2.lastBet);
    lateralSpaces();
    //printf("%c\n", 179);
    printf("│\n");
}

void printPot(int pot){
    ////printf("%c\n", 179);
    printf("│");
    spaces(40);
    printf("Pot: %d", pot);
    spaces(45 - numDigits(pot));
    //printf("%c\n", 179);
    printf("│\n");
}

void printTable(player players[], card cards[], int pot){
    topBorder();
    centralCard(players[3]);
    printCentralPlayer(players[3], 4);
    cardsLateral(players[2], players[4]);
    printLateralPlayers(players[2], 3, players[4], 5);

    flopTurnRiver(cards);
    printPot(pot);

    cardsLateral(players[1], players[5]);
    printLateralPlayers(players[1], 2, players[5], 6);

    printCentralPlayer(players[0], 1);
    float prob = 0;
    if(cards[0].value == ' '){
        prob = players[0].preFlopProb;
    }else if(cards[3].value == ' '){
        prob = players[0].flopToTurnProb;
    }else if(cards[4].value == ' '){
        prob = players[0].turnToRiverProb;
    }else{
        prob = players[0].riverToShowDownProb;
    }
    centralCardWithProb(players[0], prob);
    bottomBorder();
}

int main(){
    card c1;
    c1.value = 'K';
    c1.naipe = 'P';
    card c2;
    c2.value = 'Q';
    c2.naipe = 'C';
    card c3;
    c3.value = 'J';
    c3.naipe = 'O';
    card c4;
    c4.value = ' ';
    c4.naipe = ' ';
    card c5;
    c5.value = ' ';
    c5.naipe = ' ';

    player p1;
    p1.hand[0] = c1;
    p1.hand[1] = c2;
    p1.chips = 50000;
    p1.role = 'C';
    p1.preFlopProb = 3;
    p1.flopToTurnProb = 0;
    p1.turnToRiverProb = 7.2;
    p1.riverToShowDownProb = 2.78;
    p1.active = true;
    p1.lastBet = 0;
    p1.showCards = false;
    
    player p2;
    p2.hand[0] = c3;
    p2.hand[1] = c4;
    p2.chips = 500;
    p2.role = 'C';
    p2.active = false;
    p2.lastBet = 0;
    p2.showCards = false;

    player p3;
    p3.hand[0] = c1;
    p3.hand[1] = c1;
    p3.chips = 500;
    p3.role = 'D';
    p3.active = true;
    p3.lastBet = 951738;
    p3.showCards = true;

    player p4;
    p4.hand[0] = c3;
    p4.hand[1] = c1;
    p4.chips = 50000;
    p4.role = 'B';
    p4.active = false;
    p4.lastBet = 500;
    p4.lastBet = 8000;
    p4.showCards = false;

    player p5;
    p5.hand[0] = c2;
    p5.hand[1] = c2;
    p5.chips = 500000;
    p5.role = 'S';
    p5.active = false;
    p5.lastBet = 0;
    p5.showCards = true;

    player p6;
    p6.hand[0] = c1;
    p6.hand[1] = c2;
    p6.chips = 500;
    p6.role = 'C';
    p6.active = false;
    p6.lastBet = 0;
    p6.showCards = false;

    player players[] = {p1, p2, p3, p4, p5, p6};
    card cards[] = {c1, c2, c3, c4, c5};

    printTable(players, cards, 50000);//p1, p2, p3, p4, p5, p6, c1, c2, c3, c4, c5, 500000);
}

/*
    ╔══════════════════════════════════════════════════════════════════════════════════════════╗
    ║                                                                                          ║
    ║                                                                                          ║
    ║   ╔═══╗╔═══╗                                                                             ║
    ║   ║8 C║║3 P║                                                                             ║
    ║   ║   ║║   ║                                                                             ║
    ║   ╚═══╝╚═══╝                                                                             ║
    ║                                                                                          ║
    ║   ┌───┐┌───┐                                                                             ║
    ║   │8 O││3 E│                                                                             ║
    ║   │   ││   │                                                                             ║
    ║   └───┘└───┘                                                                             ║
    ║                                                                                          ║
    ╚══════════════════════════════════════════════════════════════════════════════════════════╝ 
   
    ┌──────────────────────────────────────────────────────────────────────────────────────────┐
    │                                                                                          │
    │                                                                                          │
    │   ╔═══╗╔═══╗                                                                             │
    │   ║8 C║║3 P║                                                                             │
    │   ║   ║║   ║                                                                             │
    │   ╚═══╝╚═══╝                                                                             │
    │                                                                                          │
    │   ┌───┐┌───┐                                                                             │
    │   │8 O││3 E│                                                                             │
    │   │   ││   │                                                                             │
    │   └───┘└───┘                                                                             │
    ├──────────┐                                                                                        │
    │ WIN: 00% │                                                                                        │
    └──────────┴───────────────────────────────────────────────────────────────────────────────┘
   193 ┴
   195 ├
    │ = 179
    └ = 192
    ─ = 196
    ┐ = 191
    ┘ = 217
    ┌ = 218
    ╔ = 201
	═ = 205
	╗ = 187
	║ = 186
	╝ = 188
	╚ = 200
    ┌───┐┌───┐
    │8  ││3  │
    │   ││   │
    └───┘└───┘
    ┌───┐┌───┐
    │8  ││3  │
    │O  ││E  │
    └───┘└───┘
    ╔═══╗╔═══╗
    ║8  ║║3  ║
    ║   ║║   ║
    ╚═══╝╚═══╝
*/

File: haskell/poker/TableOutput.hs
module TableOutput(
    printTable
)where
    
import Model
import Text.Printf

printNTimes :: String -> Int -> IO()
printNTimes string num | num == 1 = putStr(string)
                 | num == 0 = putStr("")
                 | num < 0 = putStr("")
                 | otherwise = do
                    putStr(string)
                    printNTimes string (num - 1)

numberOfDigits :: Float -> Float -> Float -> Float
numberOfDigits num digits base 
    | division >= 1 = numberOfDigits num (digits + 1) (base * 10)
    | otherwise = digits
    where division = (num / base)


numDigits :: Float -> Float
numDigits num = numberOfDigits num 1 10


topBorder :: IO()
topBorder = do
    putStr("┌")
    printNTimes "─" 90
    putStr("┐\n")

bottomBorder :: IO()
bottomBorder = do
    putStr("└")
    printNTimes "─" 12
    putStr("┴")
    printNTimes "─" 77
    putStr("┘\n")

lateralBorder :: IO()
lateralBorder = do
    putStr("│")
    printNTimes " " 90
    putStr("│\n")

nLateralBorders :: Int -> IO()
nLateralBorders n | n == 1 = lateralBorder
                  | otherwise = do 
                        lateralBorder
                        nLateralBorders (n - 1)

spaces :: Int -> IO()
spaces n  | n == 1 = putStr(" ")
          | n == 0 = putStr("")
          | otherwise = do 
                putStr(" ")
                spaces (n - 1)

centralSpaces :: IO()
centralSpaces = spaces 64

centralCardSpaces :: IO()
centralCardSpaces = spaces(40)

lateralSpaces :: IO()
lateralSpaces = spaces 3

cardTop :: IO()
cardTop = do
    putStr("┌")
    printNTimes "─" 3
    putStr("┐")

nCardTops :: Int -> IO()
nCardTops n | n == 1 = cardTop
            | otherwise = do
                cardTop
                nCardTops (n - 1)

cardLateral :: String -> IO()
cardLateral info = do
    putStr("│")
    putStr(formatedInfo info)
    putStr("│")

nCardLaterals :: Int -> IO()
nCardLaterals n | n == 1 = cardLateral " "
                | otherwise = do
                    cardLateral " "
                    nCardLaterals (n - 1)

formatedInfo :: String -> String
formatedInfo string | string == "T" = "10 "
                    | otherwise = string ++ "  "

cardLateralInfo :: String -> IO()
cardLateralInfo info = do
    putStr("│")
    putStr(formatedInfo info)
    putStr("│")

cardBottom :: IO()
cardBottom = do
    putStr("└")
    printNTimes "─" 3
    putStr("┘")

nCardBottoms :: Int -> IO()
nCardBottoms n | n == 1 = cardBottom
               | otherwise = do   
                    cardBottom
                    nCardBottoms (n - 1)

centralCards :: IO()
centralCards = do
    putStr("│")

    centralCardSpaces
    nCardTops 2
    centralCardSpaces

    putStr("│\n")
    
    -- Inserir Logica para mostrar ou não a carta do jogador

    putStr("│")
    centralCardSpaces
    nCardLaterals 2
    centralCardSpaces
    putStr("│\n")

    putStr("│")
    centralCardSpaces
    nCardLaterals 2
    centralCardSpaces
    putStr("│\n")

    putStr("│")
    centralCardSpaces
    nCardBottoms 2
    centralCardSpaces
    putStr("│\n")


centralCardsWithProb :: Player -> Float -> IO()
centralCardsWithProb (Player {hand = h, chips = c, active = a}) prob = do
    putStr("│")

    centralCardSpaces
    nCardTops 2
    centralCardSpaces

    putStr("│\n")
      
    putStr("│")
    centralCardSpaces
    cardLateral (value (h!!0))
    cardLateral (value (h!!1))
    centralCardSpaces
    putStr("│\n")

    putStr("├")
    printNTimes "─" 12
    putStr("┐")
    
    spaces 27
    cardLateral (naipe (h!!0))
    cardLateral (naipe (h!!1))
    centralCardSpaces
    putStr("│\n")

    putStr("│")
    putStr(" WIN: ")
    printf "%.1f" (prob :: Float) -- ++ show(100.0) ++ "%")
    putStr("%")
    spaces (3 - truncate(numDigits prob))
    putStr("│")
    spaces 27
    nCardBottoms 2
    centralCardSpaces
    putStr("│\n")

lateralCards :: IO()
lateralCards = do
    putStr("│")
    lateralSpaces
    nCardTops 2
    centralSpaces
    nCardTops 2
    lateralSpaces
    putStr("│\n")

    -- Inserir Logica para mostrar ou não a carta do jogador
    putStr("│")
    lateralSpaces
    nCardLaterals 2
    centralSpaces

    nCardLaterals 2
    lateralSpaces
    putStr("│\n")

    putStr("│")
    lateralSpaces
    nCardLaterals 2
    centralSpaces

    nCardLaterals 2
    lateralSpaces
    putStr("│\n")

    putStr("│")
    lateralSpaces
    nCardBottoms 2
    centralSpaces
    nCardBottoms 2
    lateralSpaces
    putStr("│\n")
    
flopTurnRiver :: [Card] -> IO()
flopTurnRiver cards = do
        putStr("│")
        spaces 30
        nCardTops 3
        spaces 2
        cardTop
        spaces 2
        cardTop
        spaces 31
        putStr("│\n")
    
        putStr("│")
        spaces 30
        cardLateral (value (cards !! 0))
        cardLateral (value (cards !! 1))
        cardLateral (value (cards !! 2))
        spaces 2
        cardLateral (value (cards !! 3))
        spaces 2
        cardLateral (value (cards !! 4))
        spaces 31
        putStr("│\n")
    
        putStr("│")
        spaces 30
        cardLateral (naipe (cards !! 0))
        cardLateral (naipe (cards !! 1))
        cardLateral (naipe (cards !! 2))
        spaces 2
        cardLateral (naipe (cards !! 3))
        spaces 2
        cardLateral (naipe (cards !! 4))
        spaces 31
        putStr("│\n")
    
        putStr("│")
        spaces 30
        nCardBottoms 3
        spaces 2
        cardBottom
        spaces 2
        cardBottom
        spaces 31
        putStr("│\n")

activePlayer :: Bool -> Int -> Int -> IO()
activePlayer active playerNum actualPlayer
    | (not active) = putStr("X")
    | playerNum == actualPlayer = putStr("*")
    | otherwise = putStr(" ")

role :: Int -> [Int] -> IO()
role playerNum blinds
    | playerNum == (blinds !! 0) + 1 = putStr("(D)")
    | playerNum == (blinds !! 1) + 1 = putStr("(S)")
    | playerNum == (blinds !! 2) + 1 = putStr("(B)")
    | otherwise = putStr("   ")

centralPlayer :: Int -> Player -> Int -> [Int] -> IO()
centralPlayer player (Player {hand = h, chips = c, active = a}) playing blinds = do
    putStr("│")
    spaces 39
    activePlayer a player playing
    putStr("Player " ++ show(player))
    role player blinds
    spaces 39
    putStr("│\n")

    putStr("│")
    spaces 40
    putStr("Chips: " ++ show(c))
    spaces (43 - truncate (numDigits (fromIntegral c)))
    putStr("│\n")

lateralPlayers :: Int -> Player -> Int -> Player -> Int -> [Int] -> IO()
lateralPlayers player1 (Player {hand = h1, chips = c1, active = a1})  player2 (Player {hand = h2, chips = c2, active = a2})  playing blinds = do
    putStr("│")
    spaces 2
    activePlayer a1 player1 playing
    putStr("Player " ++ show(player1))
    role player1 blinds
    spaces 62
    role player2 blinds
    putStr("Player " ++ show(player2))
    activePlayer a2 player2 playing
    spaces 2
    putStr("│\n")

    putStr("│")
    lateralSpaces
    putStr("Chips: " ++ show(c1))
    spaces (70 - truncate(numDigits (fromIntegral c1)) - truncate(numDigits (fromIntegral c2)))
    putStr("Chips: " ++ show(c2))
    lateralSpaces
    putStr("│\n")

printPot :: Int -> IO()
printPot chips = do
    putStr("│")
    spaces 40
    putStr("Pot: " ++ show(chips))
    spaces (45 - truncate(numDigits (fromIntegral chips)))
    putStr("│\n")
    
getProb :: Player -> Int -> Float
getProb player currentRound
    | currentRound == 0 = preFlopProb player
    | currentRound == 1 = flopToTurnProb player
    | currentRound == 2 = turnToRiverProb player
    | currentRound == 3 = riverToShowDownProb player
    | otherwise = 0

printTable :: GameStatus -> IO()
printTable gameStatus = do

    let cards = cardsTable gameStatus
    let players = playersTable gameStatus
    let round = currentRound gameStatus
    let potChips = pot gameStatus
    let playing = (actualPlayer gameStatus) + 1
    let dealer = dealerPosition gameStatus
    let blinds = [dealer, smallPosition gameStatus, bigPosition gameStatus]

    topBorder
    centralCards
    centralPlayer 4 (players !! 3) playing blinds
    lateralCards
    lateralPlayers 3 (players !! 2) 5 (players !! 4) playing blinds
    flopTurnRiver cards
    printPot potChips
    lateralCards
    lateralPlayers 2 (players !! 1) 6 (players !! 5) playing blinds
    centralPlayer 1 (players !! 0) playing blinds
    centralCardsWithProb (players!!0) (getProb (players!!0) round)
    bottomBorder


getValue :: Card -> String
getValue (Card {naipe = n, value = v}) = v


getNaipe :: Card -> String
getNaipe (Card {naipe = n, value = v}) = n


printPlayer :: Player -> IO()
printPlayer (Player {hand = h, chips = c, active = a}) = do
    printCard (h!!0)
    printCard (h!!1)
    putStrLn(show c)
    putStrLn(show a)
    


printCard :: Card -> IO()
printCard (Card {naipe = n, value = v}) = do
    putStrLn(n)
    putStrLn(v)


printUserProfile :: GameStatus -> IO()
printUserProfile gs = do 
   putStr "\n\n        .------..------..------..------..------..------.\n"
   putStr "        |P.--. ||E.--. ||R.--. ||F.--. ||I.--. ||L.--. |\n"
   putStr "        | \58/\92\58 || \58/\92\58 || \58()\58 || \58()\58 || \58/\92\58 || \58/\92\58 |\n"
   putStr "        | (__) || (__) || ()() || ()() || (__) || (__) |\n"
   putStr "        | '--'P|| '--'E|| '--'R|| '--'F|| '--'I|| '--'L|\n"
   putStr "        `------'`------'`------'`------'`------'`------'\n\n\n"
   putStr "PERFIS POSSÍVEIS \n\n[-] MUITO AGRESSIVO [-] AGRESSIVO [+] MUITO MODERADO [+] MODERADO\n\n\n"

   let average_prob = ((preFlopProb ((playersTable gs)!!0)) + (flopToTurnProb ((playersTable gs)!!0)) + (turnToRiverProb ((playersTable gs)!!0)) + (riverToShowDownProb ((playersTable gs)!!0))) / 4 
   let average_pot = (pot gs) `div` 4
   putStr ""
   
   {-}   float average_prob = 
        (playersTable[0].preFlopProb + playersTable[0].flopToTurnProb + 
        playersTable[0].turnToRiverProb + playersTable[0].riverToShowDownProb) / 4;
    float average_pot = POT / 4;
    // MÃO NÃO MELHOROU AO LONGO DOS TURNOS
    if (average_prob < playersTable[0].flopToTurnProb) {

        // USUARIO COM POUCAS FICHAS 
        if (playersTable[0].chips < average_pot) {
            profile += "Seu perfil é MUITO AGRESSIVO, no geral sua probabilidade\n";
            profile += "de vitória não melhorou em relação a sua probabilidade no \nFLOP ";
            profile += "e suas fichas estão abaixo da média do pote.";
        } else {
            profile += "Seu perfil é AGRESSIVO, no geral sua probabilidade\n";
            profile += "de vitória não melhorou em relação a sua probabilidade \n no FLOP,\n";
            profile += "mas suas fichas estão acima da média do pote.";
        }
    // MÃO MELHOROU AO LONGO DOS TURNOS    
    } else {
        // USUARIO COM POUCAS FICHAS 
        if (playersTable[0].chips < average_pot) {
            profile += "Seu perfil é MUITO MODERADO, no geral sua probabilidade\n"; 
            profile += "de vitória melhorou em relação a sua probabilidade no FLOP,\n";
            profile += "mas suas fichas estão abaixo da média do pote"; 
        } else {
            profile += "Seu perfil é MODERADO, no geral sua probabilidade\n"; 
            profile += "de vitória melhorou em relação a sua probabilidade no FLOP\n";  
            profile += "e suas fichas estão acima da média do pote";
        }
    }-}

main :: IO()
main = do
    let card1 = Card "O" "T"
    let card2 = Card "P" "K"
    let card3 = Card " " " "
    let player1 = Player [card1, card2] 500 True 100 90 80 70 
    let player2 = Player [card1, card2] 1800 False 100 100 100 100
    let player3 = Player [card1, card2] 0 True 100 100 100 100
    let player4 = Player [card1, card2] 23 True 100 100 100 100
    let player5 = Player [card1, card2] 65 False 100 100 100 100
    let player6 = Player [card1, card2] 1000 True 100 100 100 100
    let deck = Deck [card1, card2, card3]
    
    let gameStatus = GameStatus [card1, card2, card1, card3, card3] [player1, player2, player3, player4, player5, player6] 5 2 2 6 1 0 500 0 3 deck False
    printTable gameStatus

    
    

    